src/realm.js

```diff
@@ -1522,15 +1530,24 @@ export class Realm {
     });
   }
 
-  // Restores each PropertyBinding in the given map to the value it
-  // had when it was entered into the map and updates the map to record
-  // the value the Binding had just before the call to this method.
-  restoreProperties(modifiedProperties: void | PropertyBindings) {
+  redoProperties(modifiedProperties: void | PropertyBindings) {
     if (modifiedProperties === undefined) return;
-    modifiedProperties.forEach((desc, propertyBinding, m) => {
-      let d = propertyBinding.descriptor;
-      propertyBinding.descriptor = desc;
-      m.set(propertyBinding, d);
+    modifiedProperties.forEach((entry, propertyBinding, m) => {
+      if (this.capturePreState) {
```

Comment:
What is going on here is that the existing pre-state may be out of date because common subexpression elimination may have changed the values of some properties before a nested set of Effects is applied (via redoProperties). Without updating the pre-state in, we cannot safely undo the nested Effects without breaking the processing of the outer effects.

This is an artifact of applying and undoing effects while serializing and probably another reason for getting rid of the Generator.fromEffects logic.

I also wonder if it may be related to the failures in the React test suite.